proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,207 @@
+* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
+* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
+* Prevent LDAP injection attacks
+* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
+* Screens are accessible and friendly to screen readers
+* Screens are friendly to password managers
+
+#### Non-goals
+* A rich client (ie the use of javascript)
+* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
+* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
+* Supporting SSO integrations
+* Internationalization or localization. The CLI doesn't currently support this either.
+
+### Specification / How it Solves the Use Cases
+
+#### API Changes
+
+The supervisor must accept requests from other clients, as detailed
+in the (todo) proposal for dynamic client registration.
+When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
+IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
+The login page should display the IDP name and indicate that it belongs to Pinniped.
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The discovery metadata for LDAP/AD IDPS should indicate that they support a flow of `browser_authcode`.
+
+The state param should be augmented to include the IDP type as well as the IDP name. The type
+should be included in `UpstreamStateParamData` so that later when we get it back in the callback
+request we can tell which IDP it is referring to. This will require an update to
+`UpstreamStataParamData.FormatVersion`, which would mean that logins in progress at the time of
+upgrade would fail.
+
+The pinniped cli should default to using the cli-based password flow, but when the `--upstream-identity-provider-flow`,
+flag specifies `browser_authcode`, it will open a browser to log in
+instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
+may wish to authenticate using the pinniped cli but without access to a terminal.
+
+Here is how the login flow might work:
+1. The supervisor receives an authorization request.
+   1. If the client_id param is not "pinniped-cli", and it includes username and password via the custom headers, reject the request.
+   2. If the request does not include the custom username/password headers, assume we want to use the webpage login.
+   3. Some day we may want to allow specifying the idp name and type as request parameters, but
+      for now we do not have multiple idps.
```

Comment:
```suggestion
   3. Today the CLI specifies the IDP name and type as request parameters, but the server currently ignores these
      since the Supervisor does not allow multiple idps today. This could be enhanced in the future to use the requested IDP when the params are present, and to show another UI page to allow the end user to choose which IDP when the params are not present. This leaves room for future multiple IDP support in this flow, however, the details are outside the scope of this proposal.
```